data_structures/src/mainnet_validations.rs

```diff
@@ -16,6 +18,169 @@ pub const FIRST_HARD_FORK: Epoch = 192000;
+                {
+                    if is_bit_n_activated(v, n) {
+                        bit_counter.votes += 1;
+                    }
+                    if (epoch_to_update - bit_counter.init) % bit_counter.period == 0 {
+                        if (bit_counter.votes * 100) / bit_counter.period >= 80 {
+                            // An offset of 21 is added to ensure that the activation of the WIP is
+                            // achieved with consolidated blocks
+                            self.wip_activation
+                                .insert(bit_counter.wip.clone(), block_epoch + 21);
+                        }
+                        bit_counter.votes = 0;
+                    }
+                }
+            }
+        }
+        self.bit_tapi_counter.last_epoch = epoch_to_update;
+    }
+
+    pub fn initialize_wip_information(&mut self) -> (Epoch, HashSet<String>) {
+        // Hardcoded information about WIPs
+        let mut bits = vec![vec![]; 32];
+        let wip_0014 = BitVotesCounter {
+            votes: 0,
+            period: 26880,
+            wip: "WIP0014".to_string(),
+            init: 500000,
+            end: 850000,
+        };
+        bits[0].push(wip_0014);
+
+        // Assessment of new WIPs
+        let mut min_epoch = u32::MAX;
+        let mut old_wips = HashSet::default();
+        for (i, wips) in bits.into_iter().enumerate() {
+            for wip in wips {
+                if self.bit_tapi_counter.contains(i, &wip.wip) {
+                    old_wips.insert(wip.wip);
+                } else {
+                    if wip.init < min_epoch {
+                        min_epoch = wip.init;
+                    }
+                    self.bit_tapi_counter.insert(i, wip);
+                }
+            }
+        }
+
+        (min_epoch, old_wips)
+    }
+}
+
+/// Struct that count the positives votes of a WIP
+#[derive(Debug, Default, Clone, Serialize, Deserialize, PartialEq)]
+pub struct BitVotesCounter {
+    pub votes: u32,
+    pub period: Epoch,
+    pub wip: String,
+    pub init: Epoch,
+    pub end: Epoch,
+}
+
+#[derive(Debug, Default, Clone, Serialize, Deserialize, PartialEq)]
+pub struct BitTapiCounter {
+    info: [Vec<BitVotesCounter>; 32],
+    last_epoch: Epoch,
+    current_length: usize,
+}
+
+impl BitTapiCounter {
+    pub fn get_mut(&mut self, bit: usize, epoch: &u32) -> Option<&mut BitVotesCounter> {
+        match self.info.get_mut(bit) {
+            Some(bit_info) => {
+                for i in bit_info {
+                    if *epoch >= i.init && *epoch < i.end {
+                        return Some(i);
+                    }
+                }
+
+                None
+            }
+            None => None,
+        }
+    }
+
+    pub fn insert(&mut self, k: usize, v: BitVotesCounter) {
+        if k >= self.info.len() {
+            log::error!(
+                "Tapi Engine: This bit position ({}) is invalid. {} has not been included",
```

Comment:
What does "included" mean here?

Reply:
if you specify in the code a wip that will be in a bit bigger than 32, that wip will not be included in the TapiEngine

Reply:
Good.